feat(cli): `--concurrency` option

relm923 opened the pull request:
Closes #1973

Add `--concurrency` parameter to `cdk deploy` command to enable concurrent deployments while respecting stack dependencies. Concurrency mode will only work with `--progress events` due to the poor interaction of concurrent deployments and the progress bar rendering.

### Open Questions
- [x] How best to write automated tests around this?
Added unit and integration tests
- [x] Should other commands (ex: `destroy`) support concurrency?
Only supporting deploy command in this PR
- [x] Any other concerns with this approach as it changes a key component of `cdk`
- [x] How should this work with the `--exclusively` flag?
Only check dependencies between requested stacks


Example output:
```sh
$ yarn cdk deploy --all --require-approval "never" --concurrency 3 --progress bar
yarn run v1.22.17
warning package.json: No license field
$ cdk deploy --all --require-approval never --concurrency 3 --progress bar

✨  Synthesis time: 16.04s

⚠️ The --concurrency flag only supports --progress "events". Switching to "events".
relm-test-1
relm-test-1: deploying...
relm-test-2: deploying...
relm-test-2: creating CloudFormation changeset...
relm-test-1: creating CloudFormation changeset...
relm-test-1 | 0/3 | 4:57:19 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-1 User Initiated
relm-test-1 | 0/3 | 4:57:24 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-1 User Initiated
relm-test-1 | 0/3 | 4:57:30 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) 
relm-test-2 | 0/3 | 4:57:19 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-2 User Initiated
relm-test-2 | 0/3 | 4:57:24 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-2 User Initiated
relm-test-2 | 0/3 | 4:57:29 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) 
relm-test-2 | 0/3 | 4:57:29 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-1 | 0/3 | 4:57:30 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-1 | 0/3 | 4:57:30 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) Resource creation Initiated
relm-test-1 | 0/3 | 4:57:32 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
relm-test-1 | 1/3 | 4:57:33 PM | CREATE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-2 | 0/3 | 4:57:30 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) Resource creation Initiated
relm-test-2 | 0/3 | 4:57:31 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
relm-test-2 | 1/3 | 4:57:32 PM | CREATE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-2 | 2/3 | 4:57:41 PM | CREATE_COMPLETE      | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) 
relm-test-2 | 3/3 | 4:57:43 PM | CREATE_COMPLETE      | AWS::CloudFormation::Stack | relm-test-2 
relm-test-1 | 2/3 | 4:57:41 PM | CREATE_COMPLETE      | AWS::SNS::Topic    | TempTopic (TempTopic9C0CBD7C) 
relm-test-1 | 3/3 | 4:57:43 PM | CREATE_COMPLETE      | AWS::CloudFormation::Stack | relm-test-1 

 ✅  relm-test-1

✨  Deployment time: 28.39s

Stack ARN:
arn:aws:cloudformation:us-east-1:XXXXXXXXXXXX:stack/relm-test-1/ffcdb210-a6fd-11ec-b05c-0a02909bf4fb

✨  Total time: 44.43s


 ✅  relm-test-2

✨  Deployment time: 28.45s

Outputs:
relm-test-2.ExportsOutputRefTempTopic9C0CBD7CC97C6BE5 = arn:aws:sns:us-east-1:XXXXXXXXXXXX:relm-test-2-TempTopic9C0CBD7C-2ETJWELY3MFD
Stack ARN:
arn:aws:cloudformation:us-east-1:XXXXXXXXXXXX:stack/relm-test-2/ffbc7400-a6fd-11ec-ac95-0ea201c9d581

✨  Total time: 44.49s

relm-test-2B
relm-test-2B: deploying...
relm-test-2B: creating CloudFormation changeset...
relm-test-2B | 0/4 | 4:57:47 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-2B User Initiated
relm-test-2B | 0/4 | 4:57:58 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-2B User Initiated
relm-test-2B | 0/4 | 4:58:04 PM | CREATE_IN_PROGRESS   | AWS::IAM::Role     | TopicRole (TopicRole3526982D) 
relm-test-2B | 0/4 | 4:58:04 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-2B | 0/4 | 4:58:05 PM | CREATE_IN_PROGRESS   | AWS::IAM::Role     | TopicRole (TopicRole3526982D) Resource creation Initiated
relm-test-2B | 0/4 | 4:58:06 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
relm-test-2B | 1/4 | 4:58:07 PM | CREATE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-2B | 2/4 | 4:58:19 PM | CREATE_COMPLETE      | AWS::IAM::Role     | TopicRole (TopicRole3526982D) 
relm-test-2B | 2/4 | 4:58:22 PM | CREATE_IN_PROGRESS   | AWS::IAM::Policy   | TopicRole/DefaultPolicy (TopicRoleDefaultPolicy489E2B68) 
relm-test-2B | 2/4 | 4:58:23 PM | CREATE_IN_PROGRESS   | AWS::IAM::Policy   | TopicRole/DefaultPolicy (TopicRoleDefaultPolicy489E2B68) Resource creation Initiated
relm-test-2B | 3/4 | 4:58:36 PM | CREATE_COMPLETE      | AWS::IAM::Policy   | TopicRole/DefaultPolicy (TopicRoleDefaultPolicy489E2B68) 
relm-test-2B | 4/4 | 4:58:37 PM | CREATE_COMPLETE      | AWS::CloudFormation::Stack | relm-test-2B 

 ✅  relm-test-2B

✨  Deployment time: 54.71s

Stack ARN:
arn:aws:cloudformation:us-east-1:XXXXXXXXXXXX:stack/relm-test-2B/10b341d0-a6fe-11ec-a535-12d49b16bf9b

✨  Total time: 70.75s


✨  Done in 99.79s.
```

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

relm923 commented on packages/aws-cdk/lib/cdk-toolkit.ts:
```diff
@@ -153,11 +154,17 @@ export class CdkToolkit {
       warning('⚠️ It should only be used for development - never use it for your production Stacks!');
     }
 
+    const stacks = stackCollection.stackArtifacts;
```
This all works because `stackArtifacts` is pre-sorted based on stack dependencies

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b161f522d27c3afd01959ee00cae1e41fb53fff2
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8614ff66-c60f-4997-adf1-6998086f144c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 80581527f2d76f8ada56b1739cb75bf056449edb
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2d5490c4-ade4-46ba-a190-f2cda0051429%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: fee735dad911a63f87383a9bca7ab7dfe9912217
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dd6cce51-52f3-4352-bca2-5a054e0944a8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f41a4e2bcd3946ba2ff71479fb4ff4d33818938e
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fc42783c-6c6b-44ae-8d64-cd3754efadd3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

relm923 commented:
@rix0rrr sorry for the direct tag but this PR didn't get auto-assigned and it looks like you've been on most of the PRs in these files. Any chance you'd be able to review or point me to the best person to get feedback?

rix0rrr reviewed (dismissed):
Thanks for taking initiative on this! I really appreciate it. However, things are rarely as simple as they seem:

- I see why you made the exception for `--require-approval`... but this doesn't seem like the best way to go about things. A solution that allows for keeping the approval might toposort into levels, and do group approvals on each level before deploying them all in parallel.

That is not the biggest concern I have though, because it's something that allows for future extension. The bigger one is:

- The stack monitor. We'll have multiple stack monitor outputs all trampling over each other. The stack monitor will need to be made multi-stack aware, and have output limitations placed on them, in order to have properly formatted output.

Also, the "total time" reporting seems to have disappeared.

relm923 commented:
Thanks for the reply @rix0rrr!

> * I see why you made the exception for `--require-approval`... but this doesn't seem like the best way to go about things. A solution that allows for keeping the approval might toposort into levels, and do group approvals on each level before deploying them all in parallel.

I agree this is where we ultimately want to get to but what are your thoughts on leaving it as is for now and following up with a separate approval refactor?

> * The stack monitor. We'll have multiple stack monitor outputs all trampling over each other. The stack monitor will need to be made multi-stack aware, and have output limitations placed on them, in order to have properly formatted output.

I'll need to do some more digging around the monitors. Any suggestions around how to test the latest changes from the CLI level?

> * Also, the "total time" reporting seems to have disappeared.

Added back in the latest version. Although the calculation is a little odd as its total synth time + individual stack deploy time but that is pre-existing

relm923 commented on packages/aws-cdk/test/integ/cli/cli.integtest.ts:
```diff
@@ -142,7 +150,17 @@ integTest('deploy', withDefaultFixture(async (fixture) => {
 integTest('deploy all', withDefaultFixture(async (fixture) => {
   const arns = await fixture.cdkDeploy('test-*', { captureStderr: false });
 
-  // verify that we only deployed a single stack (there's a single ARN in the output)
```
Pre-existing incorrect comment

rix0rrr replied:
Thanks!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: fe1348142d7d2dcf23f07061bfb9fad9eaed486e
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ebd8e66b-b58b-4d11-aab0-9d6c7834101b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e60020282de80866788002e884ebc3483c67eb29
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=40af637c-842d-4896-aa28-56018ebe4b61%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a5c97609f1118db2a33c40f022a8b9631f660fb1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=adb0a61c-aef7-4f54-955c-18be873aca37%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

relm923 commented:
@rix0rrr as far as I can tell the Stack monitors work just fine concurrently but the monitor output can be confusing. I updated the logic to require `--progress events` when using concurrency to avoid this confusion. I've added an example of the logs in the description. Let me know what you think

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: d10905ef592661db28442a23c6c51ff0f4e2d7d4
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f8058b8f-5793-41ea-a403-15e2057e24bb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

relm923 commented:
@rix0rrr how do you want to proceed here? Happy to make a separate PR to address the security approvals

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 443e24aaa2f40390bb459790829c9524a30ffda0
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=00fa8ca8-12cb-41d6-a8d2-ba39ae8faf9e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 902e278d493a5fc00521a07d60eed81adff495bd
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=28a843d9-34e5-4672-a3c6-9ef23004e779%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 27c0a4bfe98fa0d75c598bbd47d8136c63595be9
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=02cae545-1d5f-47b1-8945-5e2bef2320cf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

RomainMuller commented on packages/aws-cdk/lib/cdk-toolkit.ts:
```diff
@@ -849,6 +893,14 @@ export interface DeployOptions extends WatchOptions {
+   * Maximum number of simulatenous deployments (dependency permitting) to execute.
+   * The default is '1', which executes all deployments serially.
+   *
+   * @default - 1
```
```suggestion
   * @default 1
```

RomainMuller reviewed (commented):
So I'm also a bit worried about the fact this does not appear to have any provision around the fact that deploying too much in parallel will lead to monitoring multiple CFN deployments concurrently, which will significantly increase the throttling risk on the `DescribeStackEvents` call... the quota for this call is... low.

relm923 commented:
> So I'm also a bit worried about the fact this does not appear to have any provision around the fact that deploying too much in parallel will lead to monitoring multiple CFN deployments concurrently, which will significantly increase the throttling risk on the `DescribeStackEvents` call... the quota for this call is... low.

Thoughts on exposing the polling interval on the Stack Monitor and scaling it based on `concurrency`? - https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/lib/api/util/cloudformation/stack-activity-monitor.ts#L352

Ex: `--concurrency 4` would poll every 20 sec instead of the default 5 sec

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 9f57e255ebad12bd3106dd567581fa335477e7f2
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9531e516-680b-47b5-bc4b-615deb0f41b1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

polothy commented:
Adjustable polling time would be great.  We get rate limited every once and a while.  An ENV var, CLI flag or CDK context value to adjust this polling interval would be great.

relm923 commented:
> Adjustable polling time would be great. We get rate limited every once and a while. An ENV var, CLI flag or CDK context value to adjust this polling interval would be great.

@polothy & @RomainMuller https://github.com/aws/aws-cdk/pull/19659 should make it easier to work around the rate limiting concerns

relm923 commented:
@rix0rrr how can we move this forward and get merged? This would be a significant improvement for my team and likely anyone deploying multiple stacks outside of CDK Pipelines

stevehodgkiss commented:
> So I'm also a bit worried about the fact this does not appear to have any provision around the fact that deploying too much in parallel will lead to monitoring multiple CFN deployments concurrently, which will significantly increase the throttling risk on the DescribeStackEvents call... the quota for this call is... low.

I have a pipeline running parallel deployments with `cdkToolkit.deploy` (~10 stacks each with an ECS service/task) and haven't run into any throttling issues, so I think this would be fine for the less extreme cases.

The `CloudFormation` client is configured to retry with this configuration:

https://github.com/aws/aws-cdk/blob/ac9901ada20e0bcadcae0e6f59e5c58220328714/packages/aws-cdk/lib/api/aws-auth/sdk.ts#L96-L102

It'd be great to have this functionality in CDK 👍

fwang commented:
@relm923 Thanks for all the hard work! Looking forward to trying it out.

Btw, are the assets for all stacks published all together upfront? Or are the assets for each individual stack published right before the stack is deployed?

relm923 commented:
> @relm923 Thanks for all the hard work! Looking forward to trying it out.
> 
> Btw, are the assets for all stacks published all together upfront? Or are the assets for each individual stack published right before the stack is deployed?

Assets for each stack are published immediately before the CF stack update - https://github.com/aws/aws-cdk/blob/4cb3b2bdb959ae398ffe2f8a5a927280f5d63306/packages/aws-cdk/lib/api/cloudformation-deployments.ts#L339

This is unchanged from the current logic

dcaruso-salty commented:
We can't use CDK Pipelines and would really enjoy this addition to the CDK. Our deployment times could be greatly sped up if we could deploy stacks in parallel.

laimonassutkus commented:
@rix0rrr whatever you do, drop it. This is a gold-tier pull request that requires first-class attention. Any other features can wait. Your current involvement is nothing but disrespect for @relm923.

rix0rrr commented:
Hey all! Sorry for not being able to get to this sooner. I would like to assure you all I'd like to see this landed as well, but I've just been super busy with other stuff that's also important, and this is a very involved PR.

I'll get to it as soon as I can, I promise. 🙏

laimonassutkus commented:
In the meantime:
https://github.com/Biomapas/B.AwsCdkParallel

rix0rrr commented on packages/aws-cdk/lib/cdk-toolkit.ts:
```diff
@@ -266,8 +288,30 @@ export class CdkToolkit {
+        .every((id) => !stacksAwaitingDeploy[id]);
+
+    const enqueueStackDeploys = async () => {
+      while (stacks[0] && isStackUnblocked(stacks[0])) {
```
We're leaving some concurrency on the table here by only considering the first element from the list.

If we had dependencies:

```
C -> A
D -> B
```

And the list happened to toposort as `[A, B, C, D]`, then with this implementation even if `B` finishes we still can't start `D` because we are forced to wait for `C` to start first.

This can be changed to a `for` loop -- on every iteration, consider all stacks.

----

It should probably be a `forEach`, otherwise we might run afoul of the "[closing over a loop variable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#creating_closures_in_loops_a_common_mistake)" gotcha.

relm923 replied:
Great catch! Updated to a `forEach` loop

rix0rrr commented on packages/aws-cdk/test/integ/cli/cli.integtest.ts:
```diff
@@ -143,7 +151,17 @@ integTest('deploy', withDefaultFixture(async (fixture) => {
+}));
+
+integTest('deploy all concurrently', withDefaultFixture(async (fixture) => {
+  const arns = await fixture.cdkDeploy('test-*', {
```
I'm not sure this is safe, many stacks in the integ app were designed to be deployed in a "just-so" way. Please pick a well-defined subset of like 5 stacks or so.

relm923 replied:
I was mainly basing this off the pre-existing `deploy-all` test case - https://github.com/aws/aws-cdk/blob/9f57e255ebad12bd3106dd567581fa335477e7f2/packages/aws-cdk/test/integ/cli/cli.integtest.ts#L151

Happy to pick a different set if you'd prefer

rix0rrr reviewed (commented):
All looks pretty good, after some minor changes I'm happy to merge this. I want everyone to know not concerned about this pull request itself: the code looks good and the feature makes sense. 

I'm concerned about the inevitable follow-up requests this will bring:

* Please make it work with confirmations (likelihood: high)
* Now that more requests are executing in parallel, I'm getting more throttles and my deployments are failing as a result (likelihood: high)
* Please make it work with the other progress reporter (likelihood: low-medium)

@relm923 will you be around to do follow-ups if issues come up?

-----------

One more thing that popped into my head: what happens if failures occur while a parallel deployment is happening? As best I can imagine, some red line will appear with an error, then a whole heap of SUCCESS will be printed as another stack that's deploying at the same time completes successfully, then we will print "Stack XYZ successfully deployed", and subsequently we'll exit with an error.

Can you tell me how confusing it will look when this happens?

polothy commented:
> One more thing that popped into my head: what happens if failures occur while a parallel deployment is happening? As best I can imagine, some red line will appear with an error, then a whole heap of SUCCESS will be printed as another stack that's deploying at the same time completes successfully, then we will print "Stack XYZ successfully deployed", and subsequently we'll exit with an error.

In my custom tool that does parallel deployments, I had this same problem (especially when I was doing 50 in parallel). What I ended up doing was on error:

* Stop starting any new deployments.
* Let current deployments complete and print their results.
* Once everything is done, re-print the failures at the end (EG: Re-printing failure 1 of 5).

This makes it really easy to find what failed in a wall of text.


> Now that more requests are executing in parallel, I'm getting more throttles and my deployments are failing as a result (likelihood: high)

This can be a problem.  I think the primary thing that causes throttling is the API call to check on the status of the CFN deployment. Think before this PR it was every 3 seconds which is too low for parallel deployments.  Think for parallel deployments, we can just default to higher and then you can also use AWS SDK settings for controlling retries (EG: increase the number and the wait between retries).

relm923 commented:
> All looks pretty good, after some minor changes I'm happy to merge this. I want everyone to know not concerned about this pull request itself: the code looks good and the feature makes sense.
> 
> I'm concerned about the inevitable follow-up requests this will bring:
> 
> * Please make it work with confirmations (likelihood: high)
> * Now that more requests are executing in parallel, I'm getting more throttles and my deployments are failing as a result (likelihood: high)
> * Please make it work with the other progress reporter (likelihood: low-medium)
> 
> @relm923 will you be around to do follow-ups if issues come up?

Happy to help with follow ups as they arise
 
> One more thing that popped into my head: what happens if failures occur while a parallel deployment is happening? As best I can imagine, some red line will appear with an error, then a whole heap of SUCCESS will be printed as another stack that's deploying at the same time completes successfully, then we will print "Stack XYZ successfully deployed", and subsequently we'll exit with an error.
> 
> Can you tell me how confusing it will look when this happens?

<details>
  <summary>Example with failure (colors stripped)</summary>
  
yarn run v1.22.18
warning package.json: No license field
$ cdk deploy --all --require-approval never --concurrency 3 --progress bar

✨  Synthesis time: 16.18s

⚠️ The --concurrency flag only supports --progress "events". Switching to "events".
relm-broken-1
relm-broken-1: deploying...
relm-broken-2
relm-broken-2: deploying...
relm-test-1: deploying...
relm-test-1: creating CloudFormation changeset...
relm-broken-1: creating CloudFormation changeset...
relm-broken-2: creating CloudFormation changeset...
relm-broken-2 | 0/3 | 9:12:48 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-broken-2 User Initiated
relm-broken-2 | 0/3 | 9:12:53 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-broken-2 User Initiated
relm-broken-1 | 0/3 | 9:12:48 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-broken-1 User Initiated
relm-broken-1 | 0/3 | 9:12:53 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-broken-1 User Initiated
relm-test-1 | 0/3 | 9:12:48 PM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-1 User Initiated
relm-test-1 | 0/3 | 9:12:54 PM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack | relm-test-1 User Initiated
relm-test-1 | 0/3 | 9:12:59 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 
relm-test-1 | 0/3 | 9:12:59 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-test-1 | 0/3 | 9:13:00 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) Resource creation Initiated
relm-test-1 | 0/3 | 9:13:01 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
relm-test-1 | 1/3 | 9:13:02 PM | CREATE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-broken-2 | 0/3 | 9:12:59 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 
relm-broken-2 | 0/3 | 9:12:59 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-broken-2 | 0/3 | 9:13:00 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) Resource creation Initiated
relm-broken-2 | 0/3 | 9:13:02 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation Initiated
relm-broken-2 | 1/3 | 9:13:02 PM | CREATE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-broken-1 | 0/3 | 9:13:04 PM | CREATE_IN_PROGRESS   | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-broken-1 | 0/3 | 9:13:04 PM | CREATE_IN_PROGRESS   | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 
relm-broken-1 | 0/3 | 9:13:04 PM | CREATE_FAILED        | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) TestTopic already exists in stack arn:aws:cloudformation:us-east-1:__ACCOUNT_ID__:stack/relm-broken-2/79fc1770-cc10-11ec-8279-0a02b75d1237
        new Topic (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/aws-sns/lib/topic.ts:102:22)
        \_ new BrokenStack (/Users/relm/Development/aws-cdk-test-stack/lib/broken-stack.ts:11:18)
        \_ Object.<anonymous> (/Users/relm/Development/aws-cdk-test-stack/bin/aws-cdk-test-stack.ts:12:1)
        \_ Module._compile (node:internal/modules/cjs/loader:1103:14)
        \_ Module.m._compile (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/index.ts:1455:23)
        \_ Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        \_ Object.require.extensions.<computed> [as .ts] (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/index.ts:1458:12)
        \_ Module.load (node:internal/modules/cjs/loader:981:32)
        \_ Function.Module._load (node:internal/modules/cjs/loader:822:12)
        \_ Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        \_ phase4 (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:567:12)
        \_ bootstrap (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:85:10)
        \_ main (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:54:10)
        \_ Object.<anonymous> (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:717:3)
        \_ Module._compile (node:internal/modules/cjs/loader:1103:14)
        \_ Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        \_ Module.load (node:internal/modules/cjs/loader:981:32)
        \_ Function.Module._load (node:internal/modules/cjs/loader:822:12)
        \_ Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        \_ node:internal/main/run_main_module:17:47
relm-broken-1 | 0/3 | 9:13:05 PM | CREATE_FAILED        | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) Resource creation cancelled
        new MetadataResource (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/metadata-resource.ts:22:24)
        \_ /Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/synthesis.ts:166:5
        \_ visit (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/synthesis.ts:231:5)
        \_ visit (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/synthesis.ts:227:5)
        \_ injectMetadataResources (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/synthesis.ts:157:3)
        \_ Object.synthesize (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/private/synthesis.ts:18:3)
        \_ App.synth (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/stage.ts:180:23)
        \_ process.<anonymous> (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/core/lib/app.ts:131:45)
        \_ Object.onceWrapper (node:events:646:26)
        \_ process.emit (node:events:526:28)
        \_ process.emit (node:domain:475:12)
        \_ process.emit.sharedData.processEmitHook.installedValue [as emit] (/Users/relm/Development/aws-cdk-test-stack/node_modules/@cspotcode/source-map-support/source-map-support.js:613:40)
relm-broken-1 | 0/3 | 9:13:06 PM | ROLLBACK_IN_PROGRESS | AWS::CloudFormation::Stack | relm-broken-1 The following resource(s) failed to create: [TestTopic339EC197, CDKMetadata]. Rollback requested by user.
relm-broken-1 | 1/3 | 9:13:10 PM | DELETE_COMPLETE      | AWS::CDK::Metadata | CDKMetadata/Default (CDKMetadata) 
relm-broken-1 | 2/3 | 9:13:10 PM | DELETE_COMPLETE      | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 
relm-broken-1 | 3/3 | 9:13:11 PM | ROLLBACK_COMPLETE    | AWS::CloudFormation::Stack | relm-broken-1 
relm-broken-2 | 2/3 | 9:13:11 PM | CREATE_COMPLETE      | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 
relm-broken-2 | 3/3 | 9:13:12 PM | CREATE_COMPLETE      | AWS::CloudFormation::Stack | relm-broken-2 
relm-test-1 | 2/3 | 9:13:10 PM | CREATE_COMPLETE      | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) 

Failed resources:
relm-broken-1 | 9:13:04 PM | CREATE_FAILED        | AWS::SNS::Topic    | TestTopic (TestTopic339EC197) TestTopic already exists in stack arn:aws:cloudformation:us-east-1:__ACCOUNT_ID__:stack/relm-broken-2/79fc1770-cc10-11ec-8279-0a02b75d1237
        new Topic (/Users/relm/Development/aws-cdk/packages/aws-cdk-lib/aws-sns/lib/topic.ts:102:22)
        \_ new BrokenStack (/Users/relm/Development/aws-cdk-test-stack/lib/broken-stack.ts:11:18)
        \_ Object.<anonymous> (/Users/relm/Development/aws-cdk-test-stack/bin/aws-cdk-test-stack.ts:12:1)
        \_ Module._compile (node:internal/modules/cjs/loader:1103:14)
        \_ Module.m._compile (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/index.ts:1455:23)
        \_ Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        \_ Object.require.extensions.<computed> [as .ts] (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/index.ts:1458:12)
        \_ Module.load (node:internal/modules/cjs/loader:981:32)
        \_ Function.Module._load (node:internal/modules/cjs/loader:822:12)
        \_ Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        \_ phase4 (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:567:12)
        \_ bootstrap (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:85:10)
        \_ main (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:54:10)
        \_ Object.<anonymous> (/Users/relm/Development/aws-cdk-test-stack/node_modules/ts-node/src/bin.ts:717:3)
        \_ Module._compile (node:internal/modules/cjs/loader:1103:14)
        \_ Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        \_ Module.load (node:internal/modules/cjs/loader:981:32)
        \_ Function.Module._load (node:internal/modules/cjs/loader:822:12)
        \_ Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        \_ node:internal/main/run_main_module:17:47

 ❌  relm-broken-1 failed: Error: The stack named relm-broken-1 failed creation, it may need to be manually deleted from the AWS console: ROLLBACK_COMPLETE
    at Object.waitForStackDeploy (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/api/util/cloudformation.ts:307:11)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at prepareAndExecuteChangeSet (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/api/deploy-stack.ts:355:26)
    at deployStack (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:229:24)
    at /Users/relm/Development/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:306:13
    at run (/Users/relm/Development/aws-cdk/node_modules/p-queue/dist/index.js:163:29)

/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/api/util/cloudformation.ts:307
    throw new Error(`The stack named ${stackName} failed creation, it may need to be manually deleted from the AWS console: ${status}`);
          ^
Error: The stack named relm-broken-1 failed creation, it may need to be manually deleted from the AWS console: ROLLBACK_COMPLETE
    at Object.waitForStackDeploy (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/api/util/cloudformation.ts:307:11)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at prepareAndExecuteChangeSet (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/api/deploy-stack.ts:355:26)
    at deployStack (/Users/relm/Development/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:229:24)
    at /Users/relm/Development/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:306:13
    at run (/Users/relm/Development/aws-cdk/node_modules/p-queue/dist/index.js:163:29)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</details>

The failures are still reported [here](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/lib/api/util/cloudformation/stack-activity-monitor.ts#L512). In flight stacks will continue while remaining stacks are not updated

rittneje commented:
@relm923 Just curious, do your changes support specifying `concurrency` within cdk.json, or only via a command line flag?

relm923 commented:
> @relm923 Just curious, do your changes support specifying `concurrency` within cdk.json, or only via a command line flag?

Only via CLI currently. Happy to (or have someone else) add settings/`cdk.json` support in a follow up PR

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e15487987581b356295e0b2ea9803b5de9fa9da5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c0b6a1c0-0627-4299-a5a8-749e0e7dac13%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented:
> Happy to help with follow ups as they arise

Thanks @relm923, that's what I needed to hear 😋. It's going in.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8f1d06dcc4fe296d30e8241f361c1c8c3db02a73
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7361dc80-8599-4310-b757-4847cdec9688%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

relm923 commented:
> Unfortunately we've had to revert this: #20230
> 
> It breaks the integ tests (errors are in the other PR). Please [run them](https://github.com/aws/aws-cdk/blob/master/packages/aws-cdk/CONTRIBUTING.md#integration-tests), maybe piping the output into `less` to emulate the situation where we don't have a TTY.

Will take a look soon. I thought all the integration tests run on PRs automatically?

rix0rrr commented:
Unfortunately the term “integ test” is overloaded. We have:

* cloudFormation snapshot tests
* Tests that actually exercise the CLI on every commit

the second set is not run automatically in PRs since that has security implications: one needs to run them by hand

OperationalFallacy commented:
@rix0rrr 

>> the second set is not run automatically in PRs since that has security implications: one needs to run them by hand

Isn't it the other way around? The manual runs have security implications.

This is a good feature, yet delayed by some obscure run by hand requirements. Do you have a process to improve this?

raymondKelly commented:
@relm923 Out of interest does concurrency work down to the nested stack level or only the top level stack?
